client/consensus/src/collators/lookahead.rs

```diff
@@ -0,0 +1,540 @@
+    SO: SyncOracle + Send + Sync + Clone + 'static,
+    Proposer: ProposerInterface<Block> + Send + Sync + 'static,
+    CS: CollatorServiceInterface<Block> + Send + Sync + 'static,
+    CHP: consensus_common::ValidationCodeHashProvider<Block::Hash> + Send + 'static,
+    P: Pair + Sync + Send + 'static,
+    P::Public: AppPublic + Member + Codec,
+    P::Signature: TryFrom<Vec<u8>> + Member + Codec,
+    GOH: RetrieveAuthoritiesFromOrchestrator<
+            Block,
+            (PHash, PersistedValidationData),
+            OrchestratorAuraWorkerAuxData<P>,
+        >
+        + 'static
+        + Sync
+        + Send,
+{
+    // This is an arbitrary value which is likely guaranteed to exceed any reasonable
+    // limit, as it would correspond to 10 non-included blocks.
+    //
+    // Since we only search for parent blocks which have already been imported,
+    // we can guarantee that all imported blocks respect the unincluded segment
+    // rules specified by the parachain's runtime and thus will never be too deep.
+    const PARENT_SEARCH_DEPTH: usize = 10;
+
+    log::info!("LOOKAHEAD COLLATOR RUNNING...");
+
+    async move {
+        cumulus_client_collator::initialize_collator_subsystems(
+            &mut params.overseer_handle,
+            params.collator_key,
+            params.para_id,
+        )
+        .await;
+
+        let mut import_notifications = match params.relay_client.import_notification_stream().await
+        {
+            Ok(s) => s,
+            Err(err) => {
+                tracing::error!(
+                    target: crate::LOG_TARGET,
+                    ?err,
+                    "Failed to initialize consensus: no relay chain import notification stream"
+                );
+
+                return;
+            }
+        };
+
+        let mut collator = {
+            let params = collator_util::Params {
+                create_inherent_data_providers: params.create_inherent_data_providers.clone(),
+                block_import: params.block_import,
+                relay_client: params.relay_client.clone(),
+                keystore: params.keystore.clone(),
+                para_id: params.para_id,
+                proposer: params.proposer,
+                collator_service: params.collator_service,
+            };
+
+            collator_util::Collator::<Block, P, _, _, _, _, _>::new(params)
+        };
+
+        // If we move forward without marking the value as unchanged,
+        // the channel will assume that the value has already changed
+        // in a different step than the one we want (inside containerChainSpawner)
+        // and will not kill the lookahead collator properly if it was already running.
+        if let Some(end_lookahead_receiver) = &mut params.end_lookahead_receiver {
+            end_lookahead_receiver.mark_unchanged();
+        }
+
+        while let Some(relay_parent_header) = import_notifications.next().await {
```

Comment:
Instead of using `tokio::sync::watch::Receiver`, you can use `CancellationToken` (this is common pattern in async programming.) 

https://docs.rs/tokio-util/latest/tokio_util/sync/struct.CancellationToken.html

This does exactly what we want to achieve and also has support for child tasks which can also be cancelled once parent token is cancelled. You can use it like this:

```rust
select! {
   _ = token.cancelled() => {
                // The token was cancelled
     }, 
   parent_header = import_notifications.next() => {
       // Call main collator function
   },
}
```

Reply:
Didn't know about this, it looks almost exactly what we need here. The only problem is that we need to cancel one task, then spawn a new one, and then later cancel the new task again. I'm not sure if we can implement that using a single `CancellationToken` (we would need some kind of token factory?).

I think the current solution is good enough because it allows use to use the same `Sender` to cancel all the tasks, and both the sender and the receiver are pretty clean. But it's good to know about this primitive, I didn't find it when looking for solutions for this problem so thanks.

Reply:
You can just create a new version by doing this:
```
let token = CancellationToken::new();
```

Plus, you can technically use child tokens and if you cancel main token the main and the child tasks are cancelled. 

Problem with current solution is that it is more error prone as you need to mark changed/unchanged in two different places, basically write some wrapper code to tailer it to our specific need.  But, with cancellation token, it is exactly our use case and we do not need any wrapper there.

Reply:
We could test the new approach using `CancellationToken` yes, but I would only implement it if we get the same results (mostly in logs) as we are getting with the current way. I'm willing to implement what we decide best, what's your opinion on this @girazoki?

Reply:
I would do it in a separate PR tbh, as I expect it to not be a small refactor